### PR TITLE
History schema evolution

### PR DESCRIPTION
**Pull request by bendk**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (bendk) [inline components/places/src/history_sync/payload_evolution_tests.rs:215]: What do others think about this one?  The current code treats visits are currently considered immutable, which means if we add unknown fields, they can't change either.

Should we try to rework the code so it doesn't make that assumption?  I kind of want to rework the merging code anyways and this could be a good excuse.

Another question is if we want to allow new fields to be part of the "key".  Visits don't have a GUID, we [treat two visits the same if they have the same timestamp + transition type](https://github.com/mozilla/application-services/blob/main/components/places/src/history_sync/plan.rs#L106-L122) (except [sometimes it's just timestamp](https://github.com/mozilla/application-services/blob/main/components/places/src/storage/history.rs#L921-L928) -- one more reason that we may want to rework this code).  Would we ever want to allow new fields to be added to this key?  Right now I don't think it's worth the complexity, but I wonder what others think.

Comment (skhamis) [reply components/places/src/history_sync/payload_evolution_tests.rs:215]: One thing I'm trying to keep in mind of roundtripping this data, is really to prevent data loss until the clients that can "catch up" and actually understand the data in those fields. 

Which means -- `unknown_fields` _should_ eventually disappear as a client becomes "recent", having unknown_fields stick around even if every client can understand the schema -- seems bad (IMO). I do like the idea of reworking so that we avoid this pitfall but depends on our timelines.

To your second question, I agree given the added complexity and we don't have a spec we'd be trying to design the would-be new "keys" around -- might be better to de-prioritize that.

File: components/places/src/storage/maintenance.rs
Comment (skhamis) [inline components/places/src/storage/maintenance.rs:86]: This is a nice solution! Would be nice to have some comments of why this exists and why we need to prune (I'm assuming it's related to the immutable visits discussion). 

That being said, I'm not entirely sure how I feel of separating the maintenance of pruning unknown_fields -- wouldn't they get removed with the visits their associated with (in `prune_older_visits`) or is this used for something else? Comments here might also lessen the confusion of this.

**Review by skhamis — COMMENTED** [review]
Seems like a great start! It seems close, but the pruning of unknown_fields separate from visits gives me a little pause. Feels we might need to understand the implications of that a bit more

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (bendk) [reply components/places/src/history_sync/payload_evolution_tests.rs:215]: > Which means -- unknown_fields should eventually disappear as a client becomes "recent", having unknown_fields stick around even if every client can understand the schema -- seems bad (IMO). I do like the idea of reworking so that we avoid this pitfall but depends on our timelines.

I'm not understanding this in relation to this question.  I think we will eventually null out `unknown_fields` once we upgrade no matter what.  My question is if we should support changes to `unknown_fields` before that happens.

However, it's a really good point in general and it's making me rethink the pruning idea.

File: components/places/src/storage/maintenance.rs
Comment (bendk) [reply components/places/src/storage/maintenance.rs:86]: Sorry, I meant to add a comment here and totally forgot about it.  But now that I'm writing it down, I'm having second thoughts.  Maybe we shouldn't special case the `unknown_fields` and just continue to use the existing pruning code.

The reasoning behind the `prune_unknown_fields` is that we only sync the most recent 20 visits for each URL, so storing `unknown_fields` for visits older than that seems like a waste of space.  However, that's not really true:
  - As you point out, the client will upgrade and at that point it will want to migrate the `unknown_fields` data and start using it.
  - If you delete recent visits, then that will push older URLs into the most recent 20.  (This is related to https://github.com/mozilla/application-services/issues/5472).

Part of the motivation here is there's a lot of visits in the DB and `unknown_fields` is space inefficient.  You want to add an enum column that could be stored in as an integer, you end up storing an entire JSON string.  Maybe we can address this another way though:
  - Lets say that we are on nightly version v120 and want to add a new feature that requires a new field
  - Before starting work on that feature, we add a migration to store the new field in a column and round-trip the data.
  - The migration should be pretty trivial, so we uplift it into v119 and an ESR patch ahead of the v120 release.
  - That way users who are doing a reasonable job keeping up-to-date don't have to deal with the space penalty.

**Review by skhamis — APPROVED** [review]
Thanks for making those changes. I think it looks good! I think we'll be able to better prune the unknown_fields when the time comes to "consume" said fields.

👍 from me! But I'd wait for Linas review before merging!

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (mhammond) [reply components/places/src/history_sync/payload_evolution_tests.rs:215]: I think a visit is considered immutable because it's recording a historical event - so changing it in the future really doesn't make sense. I can't see how a different device would somehow say "this old visit I previously recorded now should have different attributes".

Similarly, I think the key question is roughly the same - it's a record of something happening at an instant - I don't see how other fields being part of the key makes sense - you'd never want to record 2 different visits of the same type at the same timestamp with different attributes - there can only really have been one visit at that instant.

If that all makes sense, maybe just update the comment to reflect that?

**Review by mhammond — APPROVED** [review]
This looks correct to me!

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (linabutler) [inline components/places/src/history_sync/payload_evolution_tests.rs:210]: Could you please add a test with deduping involved? Maybe one where we have a local record and 2 remote records, all with the same URL but different GUIDs?

AFAICR, history deduping works a little oddly in this case. On the server, only the GUID can be unique, but locally, there's an additional requirement that URLs are unique. But because of how we rotate history, and sync only a subset of your history on the first sync, it's possible for the server to end up with multiple records—each with different GUIDs—for the same URL.

(This is another reason #5472 is tricky: the tombstone only carries the GUID, and if your device deduped the page to a different GUID than the one that uploaded the tombstone, your device won't know to delete that page).

I think our deduping logic just smushes visits for all those together, and picks the first (or the last?) page GUID that it saw as the winner. I'm not sure the right semantics for unknown fields in that case—keep the fields from the record with the winning GUID, and discard the others?—but it would be good to capture them in a test, if it's not too cumbersome to add.

**Review by linabutler — APPROVED** [review]
This looks great to me, too; thanks @bendk, and apologies for the delay!

My only question is how deduping will interact with this, but more about capturing the current semantics in a test, and not trying to define new semantics for it.

I also think your approach of adding unknown fields to just the page record, instead of individual visits, is great. If we do decide to look closer at visits, I think it's worth having that discussion in the context of other changes, like supporting deletions for them, instead of trying to solve all that here.

Thank you again! 🎊

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (bendk) [reply components/places/src/history_sync/payload_evolution_tests.rs:23]: Yeah, it's basically `json!()`, but without having to type `as_object().unwrap()`.

File: components/places/src/history_sync/payload_evolution_tests.rs
Comment (bendk) [reply components/places/src/history_sync/payload_evolution_tests.rs:212]: > I think our deduping logic just smushes visits for all those together, and picks the first (or the last?) page GUID that it saw as the winner. I'm not sure the right semantics for unknown fields in that case—keep the fields from the record with the winning GUID, and discard the others?—but it would be good to capture them in a test, if it's not too cumbersome to add.

AFAICT, the current behavior is very weird:
 - We pick the GUID from the first record
 - We pick the title from the last record
 - We hope the records are in the same order next sync (which I don't think is guaranteed).

I added a test to verify the title comes from the last record.  Based on that, I made the `unknown_fields` also come from the last record.  I'm not really sure if this is right or not, what do you think?

Comment (linabutler) [reply components/places/src/history_sync/payload_evolution_tests.rs:212]: > I added[ a test to verify the title comes from the last record](https://github.com/mozilla/application-services/blob/6879a0b13b62ca23f223130871b0973d51bf8f95/components/places/src/history_sync/plan.rs#L480-L538). Based on that, and the fact that it was simpler to code, I made the `unknown_fields` also come from the last record. I'm not really sure if this is right or not, what do you think?

That sounds great to me, thank you!

**Pull request merged**
